content/advent-2017/building-medical-reports-in-go.md

```diff
@@ -0,0 +1,241 @@
+
+## Our Solution
+
+We want to rewrite it in Go.
+
+Since our main application is written in Go and it's the only application which needs to know about reports why not to move report's logic to this end? It's makes sense, but is there any good Go library out there which is easy to use and handles our known problems in a good way?
+
+The answer is [gofpdf](https://github.com/jung-kurt/gofpdf).
+
+A great library ported from PHP which have good support for everything we needed at this time. So let's plan this change.
+
+## Proof of Concept
+
+We've created a new repo to check if we could reproduce the hardest report usign gofpdf and to understand the trade-offs with this new approach.
+
+To start we need to get it as any Go library using `go get github.com/jung-kurt/gofpdf`
+
+With the library downloaded, the code to create our POC is very simple.
+
+We created a function `NewReport` and inside of it we configured how this document will looks like (e.g. font family, font color, page size, header and footer data). Let's look into it:
+
+```go
+func NewReport() *Report {
+	pdf := gofpdf.New("P", "mm", "A4", "./assets/fonts")
+	html := pdf.HTMLBasicNew()
+	encodingFunc := pdf.UnicodeTranslatorFromDescriptor("")
+
+	pdf.AddFont("ProximaNova", "", "ProximaNova-Reg-webfont.json")
+	pdf.AddFont("ProximaNova", "B", "ProximaNova-Bold-webfont.json")
+	pdf.AddFont("ProximaNova-Light", "", "ProximaNova-Light-webfont.json")
+
+	report := &Report{
+		htmlContent:  html,
+		encodingFunc: encodingFunc,
+	}
+
+	pdf.SetFont("ProximaNova", "", fontPtSize)
+	pdf.SetTextColor(75, 75, 80)
+	pdf.AliasNbPages("")
+	pdf.SetHeaderFunc(report.headerFunc)
+	pdf.SetFooterFunc(report.footerFunc)
+
+	report.document = pdf
+	return report
+}
+```
+
+Our struct `Report` just holds the configured document as you can see in the last line before actually return it.
+
+The ideia here is just to show how simple is to create your documents using this library. You could check the [docs](https://godoc.org/github.com/jung-kurt/gofpdf) to understand what each function does. Let's move one step further.
+
+The main code lives in a `HandleFunc`, which was created just to be able to see this PDF usign a browser instead of generate it in the file system.
+
+```go
+http.HandleFunc("/report", func(w http.ResponseWriter, r *http.Request) {
+    report := NewReport()
+    report.PatientHeader()
+    report.Diagnostic()
+    report.GeneList()
+    report.TechnicalResponsible()
+    report.Method()
+    report.QualityFlags()
+    report.VUS()
+    report.Comments()
+
+
+    if err := report.document.Output(w); err != nil {
+        log.Print(err)
+    }
+})
+```
+
+Each function in the `report` object (created by our `NewReport` as we saw above) is a section builder for that given name. We build the patient header, with information like name, age, etc, diagnostic section which is where we explain if the patient have a positive or negative result for that specific exam and so on.
```

Comment:
Tricky conjugation again: instead of `if the patient have a...`, try `if the patient has a...`